packages/core/core/src/assetUtils.js

```diff
@@ -0,0 +1,180 @@
+// @flow strict-local
+
+import type {
+  ASTGenerator,
+  ConfigResult,
+  File,
+  FilePath,
+  GenerateOutput,
+  Meta,
+  PackageName,
+  Stats,
+  Symbol,
+  Transformer,
+} from '@parcel/types';
+import type {Asset, Dependency, Environment} from './types';
+
+import {Readable} from 'stream';
+import {PluginLogger} from '@parcel/logger';
+import nullthrows from 'nullthrows';
+import CommittedAsset from './CommittedAsset';
+import UncommittedAsset from './UncommittedAsset';
+import loadPlugin from './loadParcelPlugin';
+import {Asset as PublicAsset} from './public/Asset';
+import PluginOptions from './public/PluginOptions';
+import {blobToStream, loadConfig, md5FromString} from '@parcel/utils';
+import {getEnvironmentHash} from './Environment';
+
+type AssetOptions = {|
+  id?: string,
+  committed?: boolean,
+  hash?: ?string,
+  idBase?: ?string,
+  filePath: FilePath,
+  type: string,
+  contentKey?: ?string,
+  mapKey?: ?string,
+  astKey?: ?string,
+  astGenerator?: ?ASTGenerator,
+  dependencies?: Map<string, Dependency>,
+  includedFiles?: Map<FilePath, File>,
+  isIsolated?: boolean,
+  isInline?: boolean,
+  isSplittable?: ?boolean,
+  isSource: boolean,
+  env: Environment,
+  meta?: Meta,
+  outputHash?: ?string,
+  pipeline?: ?string,
+  stats: Stats,
+  symbols?: Map<Symbol, Symbol>,
+  sideEffects?: boolean,
+  uniqueKey?: ?string,
+  plugin?: PackageName,
+  configPath?: FilePath,
+|};
+
+export function createAsset(options: AssetOptions): Asset {
+  let idBase = options.idBase != null ? options.idBase : options.filePath;
+  let uniqueKey = options.uniqueKey || '';
+  return {
+    id:
+      options.id != null
+        ? options.id
+        : md5FromString(
+            idBase + options.type + getEnvironmentHash(options.env) + uniqueKey,
+          ),
+    committed: options.committed ?? false,
+    hash: options.hash,
+    filePath: options.filePath,
+    isIsolated: options.isIsolated == null ? false : options.isIsolated,
+    isInline: options.isInline == null ? false : options.isInline,
+    isSplittable: options.isSplittable,
+    type: options.type,
+    contentKey: options.contentKey,
+    mapKey: options.mapKey,
+    astKey: options.astKey,
+    astGenerator: options.astGenerator,
+    dependencies: options.dependencies || new Map(),
+    includedFiles: options.includedFiles || new Map(),
+    isSource: options.isSource,
+    outputHash: options.outputHash,
+    pipeline: options.pipeline,
+    env: options.env,
+    meta: options.meta || {},
+    stats: options.stats,
+    symbols: options.symbols || new Map(),
+    sideEffects: options.sideEffects != null ? options.sideEffects : true,
+    uniqueKey: uniqueKey,
+    plugin: options.plugin,
+    configPath: options.configPath,
+  };
+}
+
+const generateResults: WeakMap<Asset, Promise<GenerateOutput>> = new WeakMap();
+
+export function generateFromAST(
```

Comment:
Is this only needed in CommittedAsset? Should we make this and the below function methods?

Reply:
Right now we generate early in `Transformation.js` if we're not scope hoisting, which has an `UncommittedAsset`.